src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -52,7 +52,55 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 .span_label(span, format!("use of possibly uninitialized {}", item_msg))
                 .emit();
         } else {
-            let msg = ""; //FIXME: add "partially " or "collaterally "
+            // FIXME Should these be moved to impl Place?
+            fn is_deep_projection<'tcx>(
+                projected_place: &Place<'tcx>,
+                place: &Place<'tcx>
+            ) -> bool {
+                let mut cursor = projected_place;
+                let mut is_proj = false;
+                loop {
+                    if cursor == place {
+                        return is_proj;
+                    }
+                    let proj = match *cursor {
+                        Place::Local(..) | Place::Static(..) => return false,
+                        Place::Projection(ref proj) => proj
+                    };
+                    is_proj = true;
+                    cursor = &proj.base;
+                }
+            }
+
+            fn root_place<'d, 'tcx>(place: &'d Place<'tcx>) -> &'d Place<'tcx> {
+                let mut cursor = place;
+                loop {
+                    let proj = match *cursor {
+                        Place::Local(..) | Place::Static(..) => return cursor,
+                        Place::Projection(ref proj) => proj,
+                    };
+                    cursor = &proj.base;
+                }
+            }
+
+            let mut msg = "";
+            let move_place = &self.move_data.move_paths[mpi].place;
+
+            // Scenarios:
+            // 1. Already moved a.b.c and trying to move a or a.b => partial move
+            // 2. Already moved a.b.c and trying to move a.b.c.d => collateral move
```

Comment:
I feel this is incorrect. What constitutes a collateral move error? 

All of these cases don't currently trigger an error. If I recall correctly only one of them does  
https://github.com/rust-lang/rust/blob/master/src/test/ui/borrowck/borrowck-box-insensitivity.rs#L96-L118

Reply:
@nikomatsakis ^

Reply:
Regarding those tests, we did indeed make the handling of `Box` more precise and I could imagine it making a lot of those tests pass now. That's ok.